### PR TITLE
Cancel LOs at BeginBlock & max JITs

### PR DESCRIPTION
**Pull request by jcompagni10**
1.) Move the expiring limit order purge to BeginBlock. Previously they were purged as part of endblock. Since they were removed before the commit stage of the block, they would never be visible to any aggregators or other services querying the dex. Since JIT orders will often represent the best-priced liquidity this lack of visibility would make the dex less favorable for routing. 

2. Enforce a maximum number of JITs per block: this creates competition for top-of-block space between market makers and arbitrageurs. 

Also, add parameters for `max_JITs_per_block` and `good_til_purge_allowance` (the amount of gas that can be used to purge GTT limit orders.) Both of these may need to be updated occasionally as market conditions change. 

More info on thought process here: https://www.notion.so/hadron/Duality-Transaction-Ordering-0a12abb340774880844d825e57be7a60#bc8cfd4af87f44d59801d6aa3ea3274a

### CODE REVIEW

**Review by pr0n00gler — DISMISSED** [review]

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9025914455

File: x/dex/types/params.pb.go [hunk 1]
```diff
@@ -28,8 +28,10 @@ const _ = proto.GoGoProtoPackageIsVersion3 // please upgrade the proto package
 
 // Params defines the parameters for the module.
 type Params struct {
-	FeeTiers           []uint64                                             `protobuf:"varint,1,rep,packed,name=fee_tiers,json=feeTiers,proto3" json:"fee_tiers,omitempty"`
-	MaxTrueTakerSpread github_com_neutron_org_neutron_v3_utils_math.PrecDec `protobuf:"bytes,2,opt,name=max_true_taker_spread,json=maxTrueTakerSpread,proto3,customtype=github.com/neutron-org/neutron/v3/utils/math.PrecDec" json:"max_true_taker_spread" yaml:"max_true_taker_spread"`
+	FeeTiers              []uint64                                             `protobuf:"varint,1,rep,packed,name=fee_tiers,json=feeTiers,proto3" json:"fee_tiers,omitempty"`
+	MaxTrueTakerSpread    github_com_neutron_org_neutron_v3_utils_math.PrecDec `protobuf:"bytes,2,opt,name=max_true_taker_spread,json=maxTrueTakerSpread,proto3,customtype=github.com/neutron-org/neutron/v3/utils/math.PrecDec" json:"max_true_taker_spread" yaml:"max_true_taker_spread"`
+	Max_JITsPerBlock      uint64                                               `protobuf:"varint,3,opt,name=max_JITs_per_block,json=maxJITsPerBlock,proto3" json:"max_JITs_per_block,omitempty"`
```
Comment (swelf19) [inline x/dex/types/params.pb.go]: Usually we serialize json field name with snake_case, let's do it the same way.

**Review by swelf19 — COMMENTED** [review]
With applying this  it is safe to remove ` EnsureBlockGasMeter` method in the app.go file, right?

**Comment by jcompagni10** [comment]
> With applying this it is safe to remove  EnsureBlockGasMeter method in the app.go file, right?

No, we still need it for limitOrder tranche key generation.  @swelf19

File: x/dex/types/params.pb.go (hunk 1 above)
Comment (jcompagni10) [reply x/dex/types/params.pb.go]: What is not snaked cased? Is the issue the capitalization of JIT?

Comment (jcompagni10) [reply x/dex/types/params.pb.go]: @swelf19

**Comment by swelf19** [comment]
> > With applying this it is safe to remove  EnsureBlockGasMeter method in the app.go file, right?
> 
> No, we still need it for limitOrder tranche key generation. @swelf19

i mean we don't need to require MaxGas to be set anymore, no? I don't see where we use it

**Comment by jcompagni10** [comment]
> i mean we don't need to require MaxGas to be set anymore, no? I don't see where we use it

Ah yes, you are correct.

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9083217334

**Review by NicholasDotSol — APPROVED** [review]

**Comment by pr0n00gler** [comment]
> https://github.com/neutron-org/neutron-tests/actions/runs/9083217334

This is definitely a wrong branch of Neutron in the test-run, please attach a correct one

Screenshot 2024-05-27 at 3 10 47 PM

**Review by pr0n00gler — CHANGES_REQUESTED** [review]
> https://github.com/neutron-org/neutron-tests/actions/runs/9083217334

This is definitely a wrong branch of Neutron in the test-run, please attach a correct one

Screenshot 2024-05-27 at 3 10 47 PM

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9258053489

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9258429370

**Review by pr0n00gler — APPROVED** [review]

**Review by swelf19 — APPROVED** [review]

**Pull request merged**